### ISSUE TITLE
Tests failing

### ISSUE THREAD

**Issue by dn0sar** [vmware/hillview#306]
Hi,

The following tests are failing when compiling the platform module:

```
[ERROR] Failures: 
[ERROR]   JavascriptTest.testDate:106 expected:<...5]
Mike,20,1990-02-0[2T23:00:00Z
John,30,2000-02-02T23:00:00Z
Tom,10,1980-02-02]T23:00:00Z
but was:<...5]
Mike,20,1990-02-0[1T23:00:00Z
John,30,2000-02-01T23:00:00Z
Tom,10,1980-02-01]T23:00:00Z

[ERROR]   JavascriptTest.testDateOutput:173 expected:<...5]
Mike,20,1990-02-0[2T23:00:00Z,2000-02-02T23:00:00Z
John,30,2000-02-02T23:00:00Z,2010-02-02T23:00:00Z
Tom,10,1980-02-02T23:00:00Z,1990-02-02]T23:00:00Z
but was:<...5]
Mike,20,1990-02-0[1T23:00:00Z,2000-02-01T23:00:00Z
John,30,2000-02-01T23:00:00Z,2010-02-01T23:00:00Z
Tom,10,1980-02-01T23:00:00Z,1990-02-01]T23:00:00Z
```

If I disable the tests everything compiles.

This is some info about the system:

OS: Ubuntu 16.04.4 LTS (GNU/Linux 4.4.0-116-generic x86_64)
java version “1.8.0_161"
node --version -> v4.2.6

**Comment by mbudiu-vmw** [comment]
This is puzzling. The programs all create dates which should be February 2, so I don't know why the results do not match. Can you please run the following simpler test on your machine as well?

```
 @Test
    public void testJSDate() throws ScriptException {
        ScriptEngineManager factory = new ScriptEngineManager();
        ScriptEngine engine = factory.getEngineByName("nashorn");
        Object obj = engine.eval("new Date(2010, 1, 2);");
        ScriptObjectMirror jsDate = (ScriptObjectMirror)obj;
        double timestampLocal = (double)jsDate.callMember("getTime");
        Instant instant = Converters.toDate(timestampLocal);
        String someDate = "2010-02-02";
        DateParsing parsing = new DateParsing(someDate);
        Instant expected = parsing.parse(someDate);
        Assert.assertEquals(expected, instant);
    }
```

**Comment by mbudiu-vmw** [comment]
@dn0sar : have you managed to run the system?

**Comment by dn0sar** [comment]
I managed to run the system by disabling all tests.
Today I fiddled a bit with the tests, the two that I put were not the only ones failing, this is the complete error message:

```
[ERROR] Failures: 
[ERROR]   JavascriptTest.testDate:122 expected:<...ble[3x15]
Mike,20,19[90-01-01T23:00:00Z
John,30,2000-01-01T23:00:00Z
Tom,10,1980-01-0]1T23:00:00Z
> but was:<...ble[3x15]
Mike,20,19[89-12-31T23:00:00Z
John,30,1999-12-31T23:00:00Z
Tom,10,1979-12-3]1T23:00:00Z
>
[ERROR]   JavascriptTest.testDateOutput:189 expected:<...ble[4x15]
Mike,20,19[90-01-01T23:00:00Z,2000-01-01T23:00:00Z
John,30,2000-01-01T23:00:00Z,2010-01-01T23:00:00Z
Tom,10,1980-01-01T23:00:00Z,1990-01-0]1T23:00:00Z
> but was:<...ble[4x15]
Mike,20,19[89-12-31T23:00:00Z,1999-12-31T23:00:00Z
John,30,1999-12-31T23:00:00Z,2009-12-31T23:00:00Z
Tom,10,1979-12-31T23:00:00Z,1989-12-3]1T23:00:00Z
>
[ERROR] Errors: 
[ERROR]   CorrelationTest.testBigCorrelation:74 » Runtime java.lang.Exception: java.lang...
[ERROR]   CorrelationTest.testCorrelation:46 » UnsatisfiedLink org.jblas.NativeBlas.dgem...
[ERROR]   HeapTopKTest.testHeapTopKTimed:47 » ArrayIndexOutOfBounds 624
[ERROR]   LAMPMapTest.testBlobs:75->testLAMPMap:60 » UnsatisfiedLink org.jblas.NativeBla...
[ERROR]   LinearProjectionTest.testDatasetProjection:82 » UnsatisfiedLink org.jblas.Nati...
[ERROR]   LinearProjectionTest.testProjection:63 » UnsatisfiedLink org.jblas.NativeBlas....
[ERROR]   ListComparatorTest.ListComparatorTestOne:74 » ArrayIndexOutOfBounds 624
[ERROR]   MatrixTest.testSimpleMatrix:48 » UnsatisfiedLink org.jblas.NativeBlas.dgemm(CC...
[ERROR]   MonoidTopKTest.MonoidTopKTest0:56 » ArrayIndexOutOfBounds 624
[ERROR]   PCATest.testLinearDataset:45 » Runtime java.lang.Exception: java.lang.Unsatisf...
[ERROR]   PercentileTest.plotPercentile:39 » ArrayIndexOutOfBounds 624
[INFO] 
[ERROR] Tests run: 170, Failures: 2, Errors: 11, Skipped: 0
```

So the extra test that you added executes successfully. I also tried disabling only the tests that appear to be producing failures or errors, and without them `mvn install` executes successfully (dah). I don't know if these tests are signalling some other problem, but so far hillview seems to be running smoothly even if these tests fail..

**Comment by mbudiu-vmw** [comment]
Interesting.
The tests that depend on jblas fail because apparently we need an extra dependency for jblas: libgfortran3. I will add this to the install-dependences.sh script. You can try to install it yourself too to see whether it solves these errors `sudo apt-get install libgfortran3`

Some of the out-of-bounds tests seem to fail because they are being run in parallel and they actually use shared data structures. I will have to fix that.

I find it very puzzling that your date test fails. I will try to guess what is happening.

**Comment by dn0sar** [comment]
Mmh it's probably something else. `libgfortran3` is already installed, version = `6.3.0-18+deb9u1`

**Comment by mbudiu-vmw** [comment]
If your maven repository is some place else, please try to find where jblas is installed; for me this is the relative path to my home directory.

**Comment by dn0sar** [comment]
Ok that's weird, it complains that `libgfortran3` is not there..

```
-- org.jblas INFO jblas version is 1.2.4
Simple benchmark for jblas

Running sanity benchmarks.

checking vector addition... ok
-- org.jblas CONFIG BLAS native library not found in path. Copying native library from the archive. Consider installing the library somewhere in the path (for Windows: PATH, for Linux: LD_LIBRARY_PATH).
-- org.jblas CONFIG ArchFlavor native library not found in path. Copying native library libjblas_arch_flavor from the archive. Consider installing the library somewhere in the path (for Windows: PATH, for Linux: LD_LIBRARY_PATH).
-- org.jblas CONFIG Loading libjblas_arch_flavor.so from /lib/static/Linux/amd64/, copying to libjblas_arch_flavor.so.
-- org.jblas ERROR Couldn't load copied link file: java.lang.UnsatisfiedLinkError: /tmp/jblas2809328084115697964/libjblas_arch_flavor.so: /tmp/jblas2809328084115697964/libjblas_arch_flavor.so: failed to map segment from shared object.

On Linux 64bit, you need additional support libraries.
You need to install libgfortran3.

For example for debian or Ubuntu, type "sudo apt-get install libgfortran3"

For more information, see https://github.com/mikiobraun/jblas/wiki/Missing-Libraries
Exception in thread "main" java.lang.UnsatisfiedLinkError: org.jblas.NativeBlas.dgemm(CCIIID[DII[DIID[DII)V
	at org.jblas.NativeBlas.dgemm(Native Method)
	at org.jblas.SimpleBlas.gemm(SimpleBlas.java:247)
	at org.jblas.DoubleMatrix.mmuli(DoubleMatrix.java:1781)
	at org.jblas.DoubleMatrix.mmul(DoubleMatrix.java:3138)
	at org.jblas.util.SanityChecks.checkMatrixMultiplication(SanityChecks.java:91)
	at org.jblas.util.SanityChecks.main(SanityChecks.java:182)
	at org.jblas.benchmark.Main.main(Main.java:114)
-- org.jblas INFO Deleting /tmp/jblas2809328084115697964/libjblas_arch_flavor.so
-- org.jblas INFO Deleting /tmp/jblas2809328084115697964
```

**Comment by mbudiu-vmw** [comment]
Is libgfortran in the LD_LIBRARY_PATH?
Normally when you install it using apt-get that should be true...

**Comment by mbudiu-vmw** [comment]
Another issue is that we need libgfortran to exist on the servers, but our installation script does not do that.
I have to tweak the ansible installation to handle that.

**Comment by mbudiu-vmw** [comment]
However, only computing the PCA requires this package.
Still, on your machine some tests will fail without this package.

**Comment by dn0sar** [comment]
Nevermind, I think I know the problem. The last server that I'm using has custom versions of every package, I tried on another server that is less restricted and I only see the two fatal errors there.

**Comment by mbudiu-vmw** [comment]
Some errors are probably due to having many cores and running lots of tests in parallel.
I hope to have fixed these, but I haven't pushed the changes yet.
Still investigating the JavaScript date error.

**Comment by dn0sar** [comment]
Thanks a lot for the help :)

**Comment by mbudiu-vmw** [comment]
Maybe you can also try this modified test in `JavaScriptTest.java`:

```
   @Test
    public void testDate() {
        ITable table = ToCatMapTest.tableWithStringColumn();
        LocalDataSet<ITable> lds = new LocalDataSet<ITable>(table);
        ColumnDescription outCol = new ColumnDescription("Date", ContentsKind.Date);
        String function = "function map(row) { return new Date(1970 + row['Age'], 0, 1); }";
        CreateColumnJSMap map = new CreateColumnJSMap(function, table.getSchema(), outCol);
        IDataSet<ITable> mapped = lds.blockingMap(map);
        ITable outTable = ((LocalDataSet<ITable>)mapped).data;

        ColumnAndConverter dateColumn = outTable.getLoadedColumn("Date");
        Instant instant = dateColumn.getDate(0);
        String expectedDate = "1990-01-01";
        DateParsing simple = new DateParsing(expectedDate);
        Instant expected = simple.parse(expectedDate);
        Assert.assertEquals(expected, instant);

        String data = outTable.toLongString(3);
        String someDate = "1990-01-01";
        DateParsing parsing = new DateParsing(someDate);
        Instant someInstant = parsing.parse(someDate);
        String s = someInstant.toString();
        String suffix = s.substring(s.indexOf('T'));
        Assert.assertEquals("Table[3x15]\n" +
                "Mike,20,1990-01-01" + suffix + "\n" +
                "John,30,2000-01-01" + suffix + "\n" +
                "Tom,10,1980-01-01" + suffix + "\n", data);
    }
```

**Comment by dn0sar** [comment]
It gives the following error:

```
[ERROR] Failures: 
[ERROR]   JavascriptTest.testDate:152 expected:<...ble[3x15]
Mike,20,19[90-01-01T23:00:00Z
John,30,2000-01-01T23:00:00Z
Tom,10,1980-01-0]1T23:00:00Z
> but was:<...ble[3x15]
Mike,20,19[89-12-31T23:00:00Z
John,30,1999-12-31T23:00:00Z
Tom,10,1979-12-3]1T23:00:00Z
>

```

**Comment by mbudiu-vmw** [comment]
This is very useful; it looks like the date is correctly computed, but when it is converted to a string it comes out wrong.

**Comment by mbudiu-vmw** [comment]
I hope that I have fixed all failing tests.
It was a time-zone problem really.

**Comment by dn0sar** [comment]
Yep, they are all passing now :)

**Comment by dn0sar** [comment]
Ps. The problem with jblas was that /tmp was not mounted with exec permissions.
`sudo mount /tmp -o remount,exec` solved the problem for the remaining failing tests
